scoutcloud/scoutcloud/src/logic/users/auth.rs

```diff
@@ -1,64 +1,144 @@
-    let token_value = headers
-        .get(AUTH_TOKEN_NAME)
-        .or(headers.get(&AUTH_TOKEN_NAME.to_uppercase()))
-        .ok_or_else(|| AuthError::NoToken)?;
+#[derive(Clone, Debug)]
+pub struct UserToken {
+    pub user: users::Model,
+    pub token: auth_tokens::Model,
+}
 
-    let (token, user) = auth_tokens::Entity::find()
-        .find_also_related(users::Entity)
-        .filter(auth_tokens::Column::Deleted.eq(false))
-        .filter(auth_tokens::Column::TokenValue.eq(token_value))
-        .one(db)
-        .await
-        .context("fetching token")?
-        .ok_or_else(|| AuthError::TokenNotFound)?;
-    let user = user.ok_or_else(|| anyhow::anyhow!("token doesn't have user_id"))?;
+impl UserToken {
+    pub async fn try_from_http_headers<C>(db: &C, headers: &HeaderMap) -> Result<Self, AuthError>
+    where
+        C: ConnectionTrait,
+    {
+        let token_value = headers
+            .get(AUTH_TOKEN_NAME)
+            .or(headers.get(&AUTH_TOKEN_NAME.to_uppercase()))
+            .ok_or_else(|| AuthError::NoToken)?
+            .to_str()
+            .map_err(|e| anyhow::anyhow!(e))?;
 
-    Ok((token, user))
-}
+        Self::try_from_token_value(db, token_value).await
+    }
 
-pub async fn create_token(
-    db: &DatabaseConnection,
-    user_id: i32,
-) -> Result<auth_tokens::Model, DbErr> {
-    let token = auth_tokens::ActiveModel {
-        user_id: Set(user_id),
-        ..Default::default()
+    pub async fn try_from_token_value<C>(
+        db: &C,
+        token_value: impl Into<String>,
+    ) -> Result<Self, AuthError>
+    where
+        C: ConnectionTrait,
+    {
+        let (token, user) = auth_tokens::Entity::find()
+            .find_also_related(users::Entity)
+            .filter(auth_tokens::Column::Deleted.eq(false))
+            .filter(uuid_eq!(
+                auth_tokens::Column::TokenValue,
+                token_value.into()
+            ))
+            .one(db)
+            .await?
+            .ok_or_else(|| AuthError::TokenNotFound)?;
+        let user = user.ok_or_else(|| anyhow::anyhow!("token doesn't have user_id"))?;
+        Ok(Self { user, token })
     }
-    .insert(db)
-    .await?;
-    Ok(token)
-}
 
-pub async fn delete_token(db: &DatabaseConnection, token_id: i32) -> Result<(), DbErr> {
-    auth_tokens::Entity::update_many()
-        .filter(auth_tokens::Column::Id.eq(token_id))
-        .col_expr(auth_tokens::Column::Deleted, Expr::value(true))
-        .exec(db)
+    pub fn has_access_to_instance(&self, instance: &Instance) -> Result<(), AuthError> {
+        if self.user.is_superuser || instance.model.creator_token_id == self.token.id {
+            Ok(())
+        } else {
+            Err(AuthError::Unauthorized("no to the instance".to_string()))
+        }
+    }
+
+    pub async fn allowed_to_create_instance<C>(&self, db: &C) -> Result<(), AuthError>
+    where
+        C: ConnectionTrait,
+    {
+        if self.user.is_superuser {
```

Comment:
Is zero balanced accounts allowed to create instances?

Reply:
for now we give super user full authentication